### PR TITLE
Browser testing support

### PR DESCRIPTION
**Pull request by pepone**
This PR brings back support to run JavaScript tests with browsers.

The setup is a bit simpler than previous setup with a static web page for each test, hosted with http-server package.

### CODE REVIEW

File: js/gulpfile.js [hunk 1]
```diff
@@ -185,44 +185,125 @@ gulp.task("test:common:generate", (cb) => {
...
+            }
+            return null;
+        },
+    };
+}
+
+gulp.task("ice:bundle", async (cb) => {
+    let bundle = await rollup({
+        input: "src/index.js",
+        plugins: [NodeMookupResolver()],
```
Comment (pepone) [inline js/gulpfile.js:206]: We use a mockup "net" module when building "ice" for the browser, this is a common strategy with WebPack too.

File: js/gulpfile.js [hunk 2]
```diff
@@ -185,44 +185,125 @@ gulp.task("test:common:generate", (cb) => {
...
+});
+
+// A rollup resolver plugin to resolve "ice" module as an external file.
+function IceResolver() {
+    return {
+        name: "ice-resolver",
+        async resolveId(source) {
+            if (source == "ice") {
+                return { id: "/ice.js", external: "absolute" };
+            }
```
Comment (pepone) [inline js/gulpfile.js:221]: We resolve `import "ice"` to an external module (not bundled) with `/ice.js` path

File: js/gulpfile.js [hunk 3]
```diff
@@ -185,44 +185,125 @@ gulp.task("test:common:generate", (cb) => {
...
             ],
             cb,
         );
     });
 
-    gulp.task(testCleanTask(name), (cb) => {
+    gulp.task(testBundleTask(name), async (cb) => {
+        let input = fs.existsSync(path.join(name, "index.js"))
+            ? path.join(name, "index.js")
+            : path.join(name, "Client.js");
```
Comment (pepone) [inline js/gulpfile.js:267]: We create a bundle for each test that is dynamically imported with `import()` the entry point is either the `index.js` file if it exists or the `Client.js`.

File: js/src/Ice/IdentityExtensions.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+import { defineStruct } from "./Struct.js";
+import { Ice as Ice_Identity } from "./Identity.js";
+const { Identity } = Ice_Identity;
+import { StreamHelpers } from "./StreamHelpers.js";
+
+// Moved here from Identity.js and Version.js to break circular dependencies
+
+defineStruct(Identity, true, true);
+export const IdentitySeqHelper = StreamHelpers.generateSeqHelper(Identity, false);
```
Comment (pepone) [inline js/src/Ice/IdentityExtensions.js:13]: There was still a circular dependency catch when building the Ice bundle. This was previously in StreamHelpers.js using a separate file avoids it.

File: js/test/Common/ControllerI.js [hunk 5]
```diff
@@ -3,26 +3,7 @@
 //
 
 import { Ice } from "ice";
-import { Test } from "Controller.js";
-
```
Comment (pepone) [inline js/test/Common/ControllerI.js:7]: We have similar static functions in TestHelper.

File: js/test/Common/ControllerI.js [hunk 6]
```diff
@@ -164,39 +152,37 @@ class ProcessControllerI extends Test.Common.BrowserProcessController {
...
         clear() {
-            this.output.val("");
+            this.output.value = "";
         }
     }
 
     const out = new Output(clientOutput);
     const serverOut = new Output(serverOutput);
 
-    const uri = new URI(document.location.href);
```
Comment (pepone) [inline js/test/Common/ControllerI.js:195]: The built-in URL can be used, no need for additional URI package.

File: js/test/Common/controller.html [hunk 7]
```diff
@@ -5,10 +5,6 @@
         <meta name="viewport" content="width=device-width, initial-scale=1.0" />
         <title>Test Suite | Ice for JavaScript</title>
         <link rel="icon" type="image/x-icon" href="/assets/favicon.ico" />
-        <link rel="stylesheet" type="text/css" href="/assets/common.css" />
```
Comment (pepone) [inline js/test/Common/controller.html:8]: We can add a simple CSS later, but I didn't want to have dependencies in a big CSS framework like foundation.

File: js/test/Ice/properties/Client.js [hunk 8]
```diff
@@ -50,29 +50,6 @@ export class Client extends TestHelper {
             for (const [key, value] of props) {
                 test(properties.getProperty(key) == value);
             }
-        } else if (typeof window !== "undefined") {
-            /* eslint-enable no-sync */
```
Comment (pepone) [inline js/test/Ice/properties/Client.js:54]: This was using jQuery ajax, maybe I should port it to use `fetch` API? But seems parse is already tested with NodeJS and shouldn't be a difference with browsers.

File: scripts/Util.py [hunk 9]
```diff
@@ -3061,10 +3061,10 @@ def __init__(self, current):
         self.url = None
         self.driver = None
         try:
-            cmd = "node -e \"require('./bin/HttpServer')()\""
+            httpServerCmd = "node node_modules/http-server/bin/http-server -p 8080 dist"
```
Comment (pepone) [inline scripts/Util.py:3064]: We need a separate https server on port 9090. For WSS testing, I will look into it in a separate PR.

File: js/src/Ice/TimerUtil.js [hunk 10]
```diff
@@ -61,13 +57,13 @@ if (typeof process != "undefined") {
         // For Browsers that support setImmediate prefer that,
         // otherwise implement it using MessageChannel
         //
-        if (isEdge() || isIE()) {
+        if (isIE()) {
```
Comment (externl) [inline js/src/Ice/TimerUtil.js:60]: We should remove IE support in a future PR.

File: js/test/Ice/properties/Client.js (hunk 8 above)
Comment (externl) [reply js/test/Ice/properties/Client.js:54]: Yeah I wouldn't bother.

**Review by externl — APPROVED** [review]
Looks good!

**Review by ReeceHumphreys — APPROVED** [review]
Looks good!

**Pull request merged**